### PR TITLE
Add overloads to shared_ptr of type for various ops

### PR DESCRIPTION
**Pull request by caleridas**
Add overloaded constructors and/or creator functions that take "shared_ptr<const type>" instead of "const type &" arguments (and hence can take over reference from caller, instead of requiring a copy operation).

In some cases, the "type copy" methods can be removed as they have no callers anymore. For others, the superseded methods will be removed once all callers are converted.

This sets the groundwork for gradually converting all call sites to use shared type references.

### CODE REVIEW

**Comment by caleridas** [comment]
This adds several overloads to existing methods, with the intent of later superseding the methods that rely on "type copy". It is too cumbersome to do all conversions in a single pass, so I think I need to go through this intermediate step.

File: jlm/rvsdg/structural-node.hpp [hunk 1]
```diff
@@ -77,20 +77,29 @@ class structural_input : public node_input
 public:
   virtual ~structural_input() noexcept;
 
-protected:
```
Comment (phate) [inline jlm/rvsdg/structural-node.hpp:80]: Why did you remove it?

File: jlm/rvsdg/structural-node.hpp [hunk 2]
```diff
@@ -77,20 +77,29 @@ class structural_input : public node_input
...
   structural_input(
       jlm::rvsdg::structural_node * node,
       jlm::rvsdg::output * origin,
       const jlm::rvsdg::port & port);
 
-public:
```
Comment (phate) [inline jlm/rvsdg/structural-node.hpp:86]: See comment above.

File: jlm/rvsdg/structural-node.hpp [hunk 3]
```diff
@@ -113,14 +122,19 @@ class structural_output : public node_output
 public:
   virtual ~structural_output() noexcept;
 
-protected:
```
Comment (phate) [inline jlm/rvsdg/structural-node.hpp:116]: Why removed?

File: jlm/rvsdg/structural-node.hpp [hunk 4]
```diff
@@ -113,14 +122,19 @@ class structural_output : public node_output
...
   structural_output(jlm::rvsdg::structural_node * node, const jlm::rvsdg::port & port);
 
-public:
```
Comment (phate) [inline jlm/rvsdg/structural-node.hpp:119]: Same here

**Review by phate — CHANGES_REQUESTED** [review]

File: jlm/rvsdg/structural-node.hpp (hunk 1 above)
Comment (caleridas) [reply jlm/rvsdg/structural-node.hpp:80]: make_shared (and make_unique) need a public constructor, otherwise they cannot construct the class. It is also essential to use make_shared<X>(...) instead of shared_ptr<X>(new X(...)) as the former will cause one allocation, the latter causes two

if the constructor needs to be made "really" private (but you generally do not need that, unless you have something like enabled_shared_from_this), it requires some construction like:

`class X {
private:
  struct private_tag {};
public:
  X(Arg1 arg1, Arg2 arg2, ..., private_tag);

  shared_ptr<X> Create(Arg1 arg1, Arg2 arg2, ...) { return make_shared<X>(arg1, arg2, ..., private_tag {}); }
`

but that is a bit much for now, also since I need to do some more cleanup later

File: jlm/llvm/ir/operators/Load.hpp [hunk 5]
```diff
@@ -193,7 +193,19 @@ class LoadVolatileOperation final : public LoadOperation
       size_t alignment)
       : LoadOperation(
           CreateOperandPorts(numMemoryStates),
-          CreateResultPorts(loadedType, numMemoryStates),
+          CreateResultPorts(
+              std::dynamic_pointer_cast<const rvsdg::valuetype>(loadedType.copy()),
```
[the remark this answers is not in the document]
Comment (caleridas) [reply jlm/llvm/ir/operators/Load.hpp]: right, changed it

File: jlm/rvsdg/structural-node.hpp (hunk 4 above)
Comment (caleridas) [reply jlm/rvsdg/structural-node.hpp:119]: as above

**Review by phate — APPROVED** [review]
